basic.go

```
@@ -1,25 +1,152 @@
+// Package auth implements Basic authentication.
 package auth
 
 import (
 	"encoding/base64"
 	"net/http"
+	"strings"
+	"time"
 
 	"github.com/codegangsta/negroni"
+	"github.com/pmylund/go-cache"
+	"golang.org/x/crypto/bcrypt"
 )
 
-// Basic returns a negroni.HandlerFunc that authenticates via Basic Auth. Writes a http.StatusUnauthorized
-// if authentication fails
-func Basic(username string, password string) negroni.HandlerFunc {
-	var siteAuth = base64.StdEncoding.EncodeToString([]byte(username + ":" + password))
-	return func(res http.ResponseWriter, req *http.Request, next http.HandlerFunc) {
-		auth := req.Header.Get("Authorization")
-		if !SecureCompare(auth, "Basic "+siteAuth) {
-			res.Header().Set("WWW-Authenticate", "Basic realm=\"Authorization Required\"")
-			http.Error(res, "Not Authorized", http.StatusUnauthorized)
+const (
+	defaultCacheExpireTime = 10 * time.Minute
+	defaultCachePurseTime  = 60 * time.Second
+	bcryptCost             = 12
+)
+
+// DataStore is a interface for retrieving hashed password by userid.
+type DataStore interface {
+	Get(userId string) (hashedPassword []byte, found bool)
+}
+
+// simpleBasic is a simple DataStore that store only one userid, hashed password pair.
+type simpleBasic struct {
+	userId         string
+	hashedPassword []byte
+}
+
+// simpleBasic.Get returns hashed password by userid.
+func (d *simpleBasic) Get(userId string) (hashedPassword []byte, found bool) {
+	if userId == d.userId {
+		return d.hashedPassword, true
+	}
+	return nil, false
+}
+
+// NewSimpleBasic returns simpleBasic builded from userid, password
+func NewSimpleBasic(userId, password string) (*simpleBasic, error) {
+	hashedPassword, err := bcrypt.GenerateFromPassword([]byte(password), bcryptCost)
```

Please return an error if err is not nil